### ISSUE TITLE
Route Handler PATCH export type error in Next.js 15

### ISSUE THREAD

**Issue by shellen** [vercel/next.js#74127]
### Link to the code that reproduces this issue

https://github.com/shellen/repro-next-route-handler

### To Reproduce

After an npm install of the attached repro, just run `npm run build` to observe a failure to compile:

Linting and checking validity of types  ..Failed to compile.

`src/app/api/test/[id]/route.ts`
`Type error: Route "src/app/api/test/[id]/route.ts" has an invalid "PATCH" export:`
  `Type "{ params: { id: string; }; }" is not a valid type for the function's second argument.`

### Current vs. Expected behavior

# Current: Build fails with type error:

`src/app/api/test/[id]/route.ts`

```
Type error: Route "src/app/api/test/[id]/route.ts" has an invalid "PATCH" export:
  Type "{ params: { id: string; }; }" is not a valid type for the function's second argument.
```
 
# Expected Results:

Build should succeed when using the documented type pattern for route handlers:

```
  export async function PATCH(
    request: NextRequest,
    { params }: { params: { id: string } }
  )
```

  Route handlers in Next.js 15 consistently throw a TypeScript error for the second argument's type, 
  specifically for PATCH methods. This occurs even when following the documented patterns and 
  trying multiple type approaches.

### Provide environment information

```bash
Operating System:
  Platform: linux
  Arch: x64
  Version: #26~22.04.1-Ubuntu SMP Thu Jul 11 22:33:04 UTC 2024
  Available memory (MB): 7930
  Available CPU cores: 2
Binaries:
  Node: 20.18.1
  npm: 10.8.2
  Yarn: 1.22.22
  pnpm: 9.15.0
Relevant Packages:
  next: 15.1.1-canary.13 // Latest available version is detected (15.1.1-canary.13).
  eslint-config-next: 15.1.1
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.7.2
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

TypeScript

### Which stage(s) are affected? (Select all that apply)

next build (local)

### Additional context

# Attempted fixes

Here are things I've tried to fix it. 

| Approach | Description | Result | Note |
|----------|-------------|---------|------|
| Custom Props Type | Used Props interface for params | Failed | Type error |
| Props Interface | Added interface for parameters | Failed | Type error |
| Next.js Types | Used NextRequest and return types | Failed | Same type error persisted |
| Base Request Type | Used standard Request type | Failed | Type error remained |
| Record Type | Used Record<string, string \| string[]> | Failed | No improvement |
| Remove Dependencies | Removed @hello-pangea/dnd | Failed | Not dependency related |
| Disable TypedRoutes | Removed from next.config | Failed | Not config related |
| RouteSegment Type | Created custom RouteSegment type | Failed | Type error persisted |
| Runtime Specification | Added nodejs runtime | Failed | No effect on type error |
| Middleware Approach | Added type validation | Failed | Same issue |
| RouteContext Type | Custom context interface | Failed | Invalid second argument |
| NodeJS Runtime + NextRequest | Combined runtime with NextRequest | Failed | Same type error |
| Handler Function Pattern | Separated business logic | Failed | ESLint any error |
| Type Assertion Pattern | Used 'as unknown as' | Failed | Type error remained |
| Custom Route Context | Created RouteContext type | Failed | Not accepted by Next.js |

This table shows systematic attempts to resolve the issue through various typing approaches and configurations, all resulting in the same core type error with the route handler's second argument.

Running React 18 or React 19 does not seem to affect it. We're using a dependency that still needs React 18 but as mentioned, rolling the react version forward didn't seem to fix either.

**Comment by jordanallred** [comment]
I am having the same issue - any remedies? I know you can parse the URL but this is not a long-term solution.

**Comment by simsketch** [comment]
Same - I've tried:

-running this in an isolated project
-uninstalling @types/react and @types/react-dom
-swapping out my eslint.config.mjs file for an .eslintrc.js file with minimal configuration
-switching node versions (16,18,20,23)
-dockerizing to rule out my local environment as being the cause

Here is an example in Docker:
```
# Use Node 20 as a base; Next.js 15 supports Node 16.8+, 
# but 20 is recommended for new projects
FROM node:20-alpine

# Create a working directory
WORKDIR /usr/src/app

# 1) Create a new Next.js 15 app with the App Router
#    Using --experimental-app ensures the "app/" directory structure
#    Or you can omit --experimental-app if you just want the standard prompts
RUN npx create-next-app@latest --experimental-app test-docker-next --use-npm --no-eslint

# Switch into the new project folder
WORKDIR /usr/src/app/test-docker-next

# 2) Ensure we're on Next.js 15 (just in case create-next-app picked up a different version)
RUN npm install next@15.1.3

# 3) Create a minimal dynamic route: app/api/guidebooks/[id]/route.ts
RUN mkdir -p app/api/guidebooks/[id] \
  && echo "import { NextRequest, NextResponse } from 'next/server';\n\nexport async function GET(req: NextRequest, { params }: { params: { id: string } }) {\n  return NextResponse.json({ id: params.id, msg: 'Docker test route' });\n}\n" \
  > app/api/guidebooks/[id]/route.ts

# 4) Build the project to see if the error occurs
RUN npm run build

# If it got here, the build worked. 
# Optionally, copy the rest of your code, or just run the server:
CMD [\"npm\", \"start\"]
```

**Comment by shellen** [comment]
If you look at my linked repro case I have found a solution but I don't know how robust it is. 

I removed the second argument from the route handler function. Extracted the dynamic route parameter id from the request URL.You can use the request.nextUrl property to access the URL and extract parameters.

Updated route.ts file:

// src/app/api/test/[id]/route.ts
import { NextRequest, NextResponse } from 'next/server';

export async function PATCH(request: NextRequest) {
  // Extract the 'id' from the request URL
  const { pathname } = request.nextUrl;
  const segments = pathname.split('/');
  const id = segments[segments.length - 1]; // Assuming 'id' is the last segment

  return NextResponse.json({ id });
}

**Comment by jordanallred** [comment]
Yeah that’s what I’ve implemented for the time being unfortunately. But this simply is not a good long-term solution and wanted to comment for visibility.

**Comment by Netail** [comment]
```ts
import { NextRequest, NextResponse } from "next/server";

export async function PATCH(request: NextRequest, { params }: { params: Promise<{ id: string }> }) {
  const { id } = await params;
  return NextResponse.json({ id });
}
```

Seems to work fine for me

**Comment by jordanallred** [comment]
> ```ts
> import { NextRequest, NextResponse } from "next/server";
> 
> export async function PATCH(request: NextRequest, { params }: { params: Promise<{ id: string }> }) {
>   const { id } = await params;
>   return NextResponse.json({ id });
> }
> ```
> 
> Seems to work fine for me.
> 
> From your git commit history it seems like you missed the `Promise<>` for params in your typing

Can confirm this works - I simply could not find the proper prototype online and every source code I have tried to drop in was not working. You are a **lifesaver**.

**Comment by Netail** [comment]
No problem, happy to help :)

**Comment by Nivnivu** [comment]
{ params }: { params: Promise<{ userId: string | string[]; fileName: string | string[] }> }
this worked for me for the second argument

**Comment by ztanner** [comment]
We've indeed switched all `params` and `searchParams` to be promise-based in v15. This should be documented [here](https://nextjs.org/docs/app/building-your-application/upgrading/version-15#params--searchparams) and we have a [codemod](https://nextjs.org/docs/app/building-your-application/upgrading/codemods#migrate-to-async-dynamic-apis) to help automatically fix occurrences in your codebase.

If there are still any examples that reference the synchronous version, they need to be updated (please let us know if you find them by filing a documentation issue! 🙏 )
